hoomd/mpcd/MPCDReversePerturbationFlow.cc

```diff
@@ -0,0 +1,339 @@
+    }
+
+/*!
+ * \param slab_distance Slab distance
+ */
+void mpcd::MPCDReversePerturbationFlow::setSlabDistance(Scalar slab_distance)
+    {
+    // check that delta is bigger than zero
+    if (slab_distance <= Scalar(0))
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: slab distance " << slab_distance
+                                  << " needs to be bigger than zero." << std::endl;
+        throw std::runtime_error("invalid slab thickness in MPCDReversePerturbationFlow");
+        }
+
+    m_slab_distance = slab_distance;
+    requestUpdateSlabs();
+    }
+
+void mpcd::MPCDReversePerturbationFlow::setSlabs()
+    {
+    const BoxDim& global_box = m_pdata->getGlobalBox();
+    const Scalar3 global_lo = global_box.getLo();
+    const Scalar3 global_hi = global_box.getHi();
+
+    const Scalar lo_pos = -m_slab_distance;
+    const Scalar hi_pos = m_slab_distance;
+    const Scalar delta_half = Scalar(0.5) * m_slab_width;
+    m_hi_pos = make_scalar2(hi_pos - delta_half, hi_pos + delta_half);
+    m_lo_pos = make_scalar2(lo_pos - delta_half, lo_pos + delta_half);
+
+    // explicit check that the two regions for velocity swapping are inside
+    // of the box entirely and don't overlap
+    if (m_hi_pos.y >= global_hi.z)
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: hi slab " << m_hi_pos.y
+                                  << "is outside of box: " << global_hi.z << "." << std::endl;
+        throw std::runtime_error("slab outside simulation box in MPCDReversePerturbationFlow");
+        }
+    if (m_lo_pos.x < global_lo.z)
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: max slab " << m_lo_pos.y
+                                  << "is outside of box: " << global_lo.z << "." << std::endl;
+        throw std::runtime_error("slab outside simulation box in MPCDReversePerturbationFlow");
+        }
+    if (m_lo_pos.y > m_hi_pos.x)
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: hi slab " << m_hi_pos.x
+                                  << " and lo slab " << m_lo_pos.y << " are overlapping."
+                                  << std::endl;
+        throw std::runtime_error("slab outside simulation box in MPCDReversePerturbationFlow");
+        }
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::MPCDReversePerturbationFlow::update(uint64_t timestep)
+    {
+    // reset momentum exchange for this step
+    m_momentum_exchange = Scalar(0);
+
+    // don't do anything if this is an empty group
+    const unsigned int n_mpcd = m_mpcd_pdata->getN();
+    if (n_mpcd == 0)
+        return;
+
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    swapPairMomentum();
+    }
+
+std::vector<std::string> mpcd::MPCDReversePerturbationFlow::getProvidedLogQuantities()
```

Comment:
We will need to look at other HOOMD 4 examples to see how to provide loggable quantities now. I think this is not the right way to do it anymore.